p2panda-rs/src/atomic/entry.rs

```diff
@@ -0,0 +1,249 @@
+#[allow(missing_copy_implementations)]
+#[derive(Error, Debug)]
+pub enum EntryError {
+    /// Links should not be set when first entry in log.
+    #[error("backlink and skiplink not valid for this sequence number")]
+    InvalidLinks,
+}
+
+impl Entry {
+    /// Validates and returns a new instance of `Entry`.
+    pub fn new(
+        log_id: &LogId,
+        message: &Message,
+        entry_hash_skiplink: Option<&Hash>,
+        entry_hash_backlink: Option<&Hash>,
+        previous_seq_num: Option<&SeqNum>,
+    ) -> Result<Self> {
+        // If it is the first entry set sequence number to 1, otherwise increment
+        let seq_num = match previous_seq_num {
+            None => SeqNum::default(),
+            Some(s) => {
+                let mut next_seq_num = s.clone();
+                next_seq_num.next().unwrap()
+            }
+        };
+
+        let entry = Self {
+            log_id: log_id.clone().to_owned(),
+            message: Some(message.clone().to_owned()),
+            entry_hash_skiplink: entry_hash_skiplink.cloned(),
+            entry_hash_backlink: entry_hash_backlink.cloned(),
+            seq_num,
+        };
+
+        entry.validate()?;
+
+        Ok(entry)
+    }
+
+    /// Signs the Bamboo entry via Ed25519 key pair and returns the hex-encoded representation.
+    pub fn sign_and_encode(&self, key_pair: &KeyPair) -> Result<EntryEncoded> {
+        // Generate message hash
+        // @TODO: Handle case where message is not set
+        let message_encoded = self.message.clone().unwrap().encode().unwrap();
+        let message_hash = message_encoded.hash();
+        let message_size = message_encoded.size();
+
+        // Convert entry links to Bamboo crate types
+        let backlink = self
+            .entry_hash_backlink
+            .clone()
+            .map(|link| link.to_yamf_hash());
+
+        let lipmaa_link = self
+            .entry_hash_skiplink
+            .clone()
+            .map(|link| link.to_yamf_hash());
+
+        // Create bamboo entry. See: https://github.com/AljoschaMeyer/bamboo#encoding for encoding
+        // details and definition of entry fields.
+        let mut entry: BambooEntry<_, &[u8]> = BambooEntry {
+            log_id: self.log_id.as_u64(),
+            is_end_of_feed: false,
+            payload_hash: message_hash.to_yamf_hash(),
+            payload_size: message_size,
+            author: PublicKey::from_bytes(&key_pair.public_key_bytes())?,
+            seq_num: self.seq_num.as_u64(),
+            backlink,
+            lipmaa_link,
+            sig: None,
+        };
+
+        let mut entry_bytes = [0u8; MAX_ENTRY_SIZE];
+        let unsigned_entry_size = entry.encode(&mut entry_bytes).unwrap();
+        let signature = key_pair.sign(&entry_bytes[..unsigned_entry_size]);
+        let sig_bytes = &signature.to_bytes()[..];
+        let signature = BambooSignature(sig_bytes.into());
+        entry.sig = Some(signature);
```

Comment:
The wasm build fails now indeed because of the unexported sign methods. I guess we need to change the methods to return the Signature as raw bytes instead, then it will work to use it in a wasm context. For bamboo we need the raw bytes anyways since the Ed25519 Signature is different than the Bamboo Signature.